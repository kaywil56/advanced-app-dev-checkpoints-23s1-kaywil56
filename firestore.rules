rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /users/{userId} {
  		allow read: if request.auth.uid != null && 
                  request.auth.uid == userId || 
                  (exists(/databases/$(database)/documents/users/$(userId)) &&
                   resource.data.group == request.resource.data.group);
  		allow write: if request.auth.uid != null && request.auth.uid == userId;
		}
    match /groups/{groupId}{
      allow read, write: if request.auth != null;
    }
    match /todos/{todoId} {
  		allow read, write: if resource.data.user_id in get(/databases/$(database)/documents/users).data.group;
		}
  }
}
