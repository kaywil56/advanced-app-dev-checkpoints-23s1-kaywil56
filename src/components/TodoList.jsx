import TodoItem from "./TodoItem";
// import firestore from "../firestore";
import { useEffect, useState } from "react";
import { addDoc, collection, deleteDoc, doc, setDoc, onSnapshot, query } from "firebase/firestore";
import firestore from "../firestore";
import { getTodos } from "../firestore.service";

const TodoList = () => {
  const [todos, setTodos] = useState([]);
  
  useEffect(() => {
    getTodos(setTodos)
  }, [])

  return (
    <ul>
      {todos.map((todo, idx) => {
        return <TodoItem key={idx} todo={todo} />;
      })}
    </ul>
  );
};

export default TodoList;
