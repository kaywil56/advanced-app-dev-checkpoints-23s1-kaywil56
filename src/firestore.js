// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { getFirestore, enableIndexedDbPersistence } from 'firebase/firestore';
// TODO: Add SDKs for Firebase products that you want to use
// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration
// For Firebase JS SDK v7.20.0 and later, measurementId is optional
const firebaseConfig = {
  apiKey: import.meta.env.VITE_API_KEY,
  authDomain: "todo-pwa-a1582.firebaseapp.com",
  projectId: "todo-pwa-a1582",
  storageBucket: "todo-pwa-a1582.appspot.com",
  messagingSenderId: "68173145197",
  appId: "1:68173145197:web:b1b87a48c76e09c8d8cb64",
  measurementId: "G-PFHCTENHF0"
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const firestore = getFirestore(app);
enableIndexedDbPersistence(firestore)
.catch((err) => {
  if (err.code == 'failed-precondition') {
      console.log("persistence can only be enabled")
  } else if (err.code == 'unimplemented') {
      console.log("The current browser does not support all of the")
  }
});

export default firestore