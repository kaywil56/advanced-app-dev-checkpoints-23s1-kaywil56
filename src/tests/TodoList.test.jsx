import { render } from "@testing-library/react";
import { describe, test, expect } from "vitest";
import TodoList from "../components/TodoList";
import { getTodos } from "../firestore.service";

vi.mock("../firestore.service", () => {
  return { getTodos: vi.fn() };
});

describe("TodoList", () => {
  test("Test TodoList matches snapshot", () => {
    const {asFragment, getByText} = render(<TodoList />)
    expect(asFragment()).toMatchSnapshot()
  });
  test("Test that onSnapShot is called on mount", () => {
    expect(getTodos).toHaveBeenCalledOnce()
  });
});
