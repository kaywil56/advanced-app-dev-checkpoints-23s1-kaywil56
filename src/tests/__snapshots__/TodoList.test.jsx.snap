// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList > Test TodoList matches snapshot 1`] = `
<DocumentFragment>
  <ul />
</DocumentFragment>
`;
